src/main/java/org/generationcp/middleware/api/germplasm/pedigree/GermplasmPedigreeServiceImpl.java

```diff
@@ -0,0 +1,346 @@
+package org.generationcp.middleware.api.germplasm.pedigree;
+
+import org.generationcp.middleware.api.germplasm.GermplasmService;
+import org.generationcp.middleware.domain.germplasm.GermplasmDto;
+import org.generationcp.middleware.hibernate.HibernateSessionProvider;
+import org.generationcp.middleware.manager.DaoFactory;
+import org.generationcp.middleware.pojos.Germplasm;
+import org.generationcp.middleware.pojos.Method;
+import org.generationcp.middleware.pojos.Name;
+import org.generationcp.middleware.pojos.Progenitor;
+import org.generationcp.middleware.util.MaxPedigreeLevelReachedException;
+
+import javax.annotation.Resource;
+import java.util.ArrayList;
+import java.util.List;
+
+public class GermplasmPedigreeServiceImpl implements GermplasmPedigreeService {
+
+	private final DaoFactory daoFactory;
+
+	@Resource
+	private GermplasmService germplasmService;
+
+	public static final int MAX_GENERATIONS_COUNT = 5;
+	private static final ThreadLocal<Integer> GENERATIONS_COUNTER = new ThreadLocal<>();
+	private static final ThreadLocal<Boolean> CALCULATE_FULL = new ThreadLocal<>();
+
+	public GermplasmPedigreeServiceImpl(final HibernateSessionProvider sessionProvider) {
+		this.daoFactory = new DaoFactory(sessionProvider);
+	}
+
+	@Override
+	public GermplasmTreeNode getGermplasmPedigreeTree(final Integer gid, Integer level, final boolean includeDerivativeLines) {
```

Comment:
Make `level` final and use another variable like:
final finalLevel = level != null ? level : rootNode.getNumberOfGenerations();

Reply:
Done